src/net/net.c

```diff
@@ -276,6 +285,60 @@ static boolean get_static_ip6_config(tuple t, struct netif *n, const char *ifnam
     return false;
 }
 
+define_closure_function(0, 2, void, net_timeout_handler,
+                 u64, expiry, u64, overruns)
+{
+    if (overruns == timer_disabled)
+        return;
+    status_handler sh;
+    spin_lock(&net.lock);
+    sh = net.ip4_complete;
+    net.ip4_complete = 0;
+    spin_unlock(&net.lock);
+    if (sh)
+        apply(sh, STATUS_OK);
+}
+
+static boolean netif_ready(struct netif *n)
+{
+    const ip_addr_t *a = netif_ip_addr4(n);
+    return !ip_addr_isany(a);
+}
+
+void ip4_when_ready(status_handler complete, timestamp timeout)
+{
+    struct netif *n;
+    if ((n = netif_get_default()) && netif_ready(n)) {
+        apply(complete, STATUS_OK);
+        return;
+    }
+    spin_lock(&net.lock);
```

Comment:
If the interface becomes ready after the above check but before we can acquire this lock, the completion will only be invoked at the timeout. We could acquire the lock before checking for netif_ready(), or alternatively we could get rid of the spinlock altogether and do `atomic_swap_64(&net.ip4_complete, 0)` to handle concurrency.